### PR TITLE
Use CodeAnalysis.Simplification to expand and reduce generated subtrees. Add tests to cover CSharpFileBuilder and CSharpSyntaxWriter classes.

### PR DESCRIPTION
**Pull request by andriipatsula**
* Utilize the `Microsoft.CodeAnalysis.CSharp.Simplifier` functionality to reduce and simplify the qualified type generation
* Extend the `IAssemblySymbolLoader` interface with the `IEnumerable<MetadataReference> GetMetadataReferences();` method to return loaded metadata references
* Implement Test for the CSharpFileBuilder and the CSharpSyntaxWriter classes. Disable 2 tests due to https://github.com/dotnet/arcade/issues/11615

Linked issues:
https://github.com/dotnet/arcade/issues/11617
https://github.com/dotnet/arcade/issues/11614

### CODE REVIEW

File: src/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs [hunk 1]
```diff
@@ -27,33 +26,57 @@ internal class CSharpFileBuilder : IAssemblySymbolWriter, IDisposable
...
+                ProjectId.CreateNewId(), VersionStamp.Create(), assembly.Name, assembly.Name, LanguageNames.CSharp));
+            project = project.AddMetadataReferences(_metadataReferences);
+
             var namespaceSymbols = EnumerateNamespaces(assembly).Where(_symbolFilter.Include);
+            var namespaceSyntaxNodes = new List<SyntaxNode>();
 
             foreach (var namespaceSymbol in namespaceSymbols)
             {
-                SyntaxNode? syntaxNode = Visit(namespaceSymbol);
+                var syntaxNode = Visit(namespaceSymbol);
```
Comment (smasher164) [inline src/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs:61]: Typically we don't use `var` unless it's obvious from the right-hand-side what the type is.

File: src/GenAPI/Microsoft.DotNet.GenAPI/CSharpSyntaxWriter.cs [hunk 2]
```diff
@@ -33,21 +35,21 @@ public CSharpSyntaxWriter(string? exceptionMessage)
         public override SyntaxNode? VisitInterfaceDeclaration(InterfaceDeclarationSyntax node)
         {
             node = (InterfaceDeclarationSyntax)base.VisitInterfaceDeclaration(node)!;
```
Comment (smasher164) [inline src/GenAPI/Microsoft.DotNet.GenAPI/CSharpSyntaxWriter.cs:37]: Using `?` instead of `!` here might be preferable, since it will early return when `VisitInterfaceDeclaration` returns a null syntax node.

Same for below.

File: src/GenAPI/Microsoft.DotNet.GenAPI/CSharpSyntaxWriter.cs [hunk 3]
```diff
@@ -81,22 +83,76 @@ public CSharpSyntaxWriter(string? exceptionMessage)
         /// <inheritdoc />
         public override SyntaxNode? VisitAccessorDeclaration(AccessorDeclarationSyntax node)
         {
-            return node.Kind() switch
+            switch (node.Kind())
             {
+                case SyntaxKind.GetAccessorDeclaration:
+                case SyntaxKind.SetAccessorDeclaration:
+                    {
+                        var accessorListSyntax = (AccessorListSyntax)node.Parent!;
```
Comment (smasher164) [inline src/GenAPI/Microsoft.DotNet.GenAPI/CSharpSyntaxWriter.cs:91]: Is it safe to disregard these null checks?

File: src/Microsoft.DotNet.ApiSymbolExtensions/AssemblySymbolLoader.cs [hunk 4]
```diff
@@ -254,6 +254,12 @@ public IEnumerable<IAssemblySymbol> LoadMatchingAssemblies(IEnumerable<IAssembly
             return matchingAssemblies;
         }
 
+        /// <inheritdoc />
+        public IEnumerable<MetadataReference> GetMetadataReferences()
```
Comment (smasher164) [inline src/Microsoft.DotNet.ApiSymbolExtensions/AssemblySymbolLoader.cs:258]: this can just be an expression based body.
```C#
public IEnumerable<MetadataReference> GetMetadataReferences() =>  _cSharpCompilation.References;
```

Comment (andriipatsula) [reply src/Microsoft.DotNet.ApiSymbolExtensions/AssemblySymbolLoader.cs]: Nice. Didn't know that interfaces could have a property member. I have this https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/interface-properties as a reference.

File: src/Microsoft.DotNet.ApiSymbolExtensions/AssemblySymbolLoader.cs [hunk 5]
```diff
@@ -255,10 +255,7 @@ public IEnumerable<IAssemblySymbol> LoadMatchingAssemblies(IEnumerable<IAssembly
         }
 
         /// <inheritdoc />
-        public IEnumerable<MetadataReference> GetMetadataReferences()
-        {
-            return _cSharpCompilation.References;
-        }
+        public IEnumerable<MetadataReference> MetadataReferences { get => _cSharpCompilation.References; }
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.ApiSymbolExtensions/AssemblySymbolLoader.cs:258]:
```suggestion
        public IEnumerable<MetadataReference> MetadataReferences => _cSharpCompilation.References;
```

File: src/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs [hunk 6]
```diff
@@ -110,12 +110,12 @@ private void Visit(IAssemblySymbol assembly)
         private SyntaxNode Visit(SyntaxNode namedTypeNode, INamedTypeSymbol namedType)
         {
             namedTypeNode = VisitInnerNamedTypes(namedTypeNode, namedType);
-
-            foreach (var member in namedType.GetMembers().Where(_symbolFilter.Include))
+
```
[the remark this answers is not in the document]
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs:113]: I'd say with the empty line the code looks cleaner. it's like a divider between two logical parts of the function

File: src/Microsoft.DotNet.ApiSymbolExtensions/AssemblySymbolLoader.cs (hunk 5 above)
Comment (andriipatsula) [reply src/Microsoft.DotNet.ApiSymbolExtensions/AssemblySymbolLoader.cs:258]: nice. thanks

File: src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+                        void Paint();
+                    }
+                    public partial interface ISurface
+                    {
+                        void Paint();
+                    }
+                
+                    public partial class SampleClass : IControl, ISurface
+                    {
+                    /*
```
Comment (smasher164) [inline src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs:299]: Do you need this comment here? It seems unrelated to being an interface implementation test.

Comment (andriipatsula) [reply src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs]: Thanks, good catch. I updated the UT to
```C#
public interface IControl
{
    void Paint();
}
public interface ISurface
{
    void Paint();
}

public class SampleClass : IControl, ISurface
{
    public void Paint()
    {
    }
}
```

File: src/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs:113]: I was referring to the added whitespace, not an empty line.

**Review by ViktorHofer — APPROVED** [review]
